inc/classes/class-imagify-files-scan.php

```diff
@@ -0,0 +1,500 @@
+<?php
+defined( 'ABSPATH' ) || die( 'Cheatin\' uh?' );
+
+/**
+ * Class handling everything that is related to "custom folders optimization".
+ *
+ * @since  1.7
+ * @author Grégory Viguier
+ */
+class Imagify_Files_Scan {
+
+	/**
+	 * Class version.
+	 *
+	 * @var    string
+	 * @since  1.7
+	 * @author Grégory Viguier
+	 */
+	const VERSION = '1.0';
+
+	/**
+	 * The single instance of the class.
+	 *
+	 * @var    object
+	 * @since  1.7
+	 * @access protected
+	 * @author Grégory Viguier
+	 */
+	protected static $_instance;
+
+	/**
+	 * The constructor.
+	 *
+	 * @since  1.7
+	 * @access protected
+	 * @author Grégory Viguier
+	 */
+	protected function __construct() {}
+
+	/**
+	 * Get the main Instance.
+	 *
+	 * @since  1.7
+	 * @access public
+	 * @author Grégory Viguier
+	 *
+	 * @return object Main instance.
+	 */
+	public static function get_instance() {
+		if ( ! isset( self::$_instance ) ) {
+			self::$_instance = new self();
+		}
+
+		return self::$_instance;
+	}
+
+	/**
+	 * Get files (optimizable by Imagify) recursively from a specific folder.
+	 *
+	 * @since  1.7
+	 * @access public
+	 * @author Grégory Viguier
+	 *
+	 * @param  string $folder An absolute path to a folder.
+	 * @return array|object   An array of absolute paths. A WP_Error object on error.
+	 */
+	public function get_files_from_folder( $folder ) {
+		// Formate and validate the folder path.
+		if ( ! is_string( $folder ) || '' === $folder || '/' === $folder || '\\' === $folder ) {
+			return new WP_Error( 'invalid_folder', __( 'Invalid folder.', 'imagify' ) );
+		}
+
+		$folder = realpath( $folder );
+
+		if ( ! $folder ) {
+			return new WP_Error( 'folder_not_exists', __( 'This folder doesn\'t exist.', 'imagify' ) );
```

Comment:
for translators, what we did on WP Rocket is to use the full negative form instead of the contraction (does not instead of doesn't), this prevent having to escape the quote too

Reply:
I also try to, but I forget sometimes, and I admit I prefer the contraction sometimes, like here: in most cases I prefer "doesn't" over "does not", but I prefer "is not" over "isn't".
But you're right, I should use "does not". Moreover, we have "This file is not a folder" and "This folder is not allowed" few lines later, so it makes sense.